src/language/pattern-types.md

```diff
@@ -0,0 +1,333 @@
+---
+title: Pattern types
+description: Pattern type reference in Dart.
+---
+
+## Cast
+
+`foo as String`
+
+Cast patterns let you insert a [type cast][] as a pattern is
+[destructuring][destructure] values:
+
+```dart
+(num, Object) record = (1, "s");
+var (i as int, s as String) = record;
+```
+
+Cast patterns will [throw][] if the value doesn't have the stated type.
+Like the [null-assert pattern](#null-assert), this lets you forcibly assert the
+expected type of some destructured value.
+
+## Constant	
+
+`123, null, 'string', math.pi, SomeClass.constant, const Thing(1, 2), const (1 + 2)`
+
+Constant patterns allow simple literals and references to named constants to be
+used directly as patterns. They match when the value is equal to the constant.
+
+Some valid constant expressions may not be valid constant patterns. In that case,
+to distinguish a constant expression from a constant pattern, prefix the expression
+with `const` and wrap it in parenthesis so it is unambiguosly an expression and
+not a constant pattern:
+
+```dart
+// List or map pattern:
+case [a, b]: ...
+
+// List or map literal:
+case const [a, b]: ...
+```
+
+## Identifier	
```

Comment:
This section is great. It does a very good job of explaining one of the more confusing parts of the feature.